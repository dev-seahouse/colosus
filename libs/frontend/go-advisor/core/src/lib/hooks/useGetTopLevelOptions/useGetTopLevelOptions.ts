import { useQuery } from '@tanstack/react-query';
import { ConnectAdvisorTopLevelOptionsApi } from '@bambu/api-client';
import type { ConnectAdvisorGetTopLevelOptionsResponseDto } from '@bambu/api-client';
import type { QueryClient } from '@tanstack/react-query';

export type GetTopLevelOptionsData =
  ConnectAdvisorGetTopLevelOptionsResponseDto;

export const getTopLevelOptionsQuery = () => ({
  queryKey: ['getTopLevelOptions'],
  queryFn: async () => {
    const connectAdvisorTopLevelOptionsApi =
      new ConnectAdvisorTopLevelOptionsApi();

    const res = await connectAdvisorTopLevelOptionsApi.getTopLevelOptions();

    return res.data;
  },
});

export const getTopLevelOptionsLoader =
  (queryClient: QueryClient) => async (): Promise<GetTopLevelOptionsData> => {
    const query = getTopLevelOptionsQuery();

    // ⬇️ return data or fetch it
    return (
      queryClient.getQueryData(query.queryKey) ??
      (await queryClient.fetchQuery(query))
    );
  };

export interface UseGetTopLevelOptionsOptions {
  initialData?: Partial<GetTopLevelOptionsData>;
}
export function useGetTopLevelOptions({
  initialData,
}: UseGetTopLevelOptionsOptions = {}) {
  return useQuery({
    ...getTopLevelOptionsQuery(),
    initialData,
  });
}

export default useGetTopLevelOptions;
