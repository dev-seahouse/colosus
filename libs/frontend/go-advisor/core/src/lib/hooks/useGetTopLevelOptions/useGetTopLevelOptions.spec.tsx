import {
  renderHook,
  queryClientWrapper,
  waitFor,
} from '@bambu/react-test-utils';

import useGetTopLevelOptions from './useGetTopLevelOptions';

describe('useGetTopLevelOptions', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetTopLevelOptions(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
