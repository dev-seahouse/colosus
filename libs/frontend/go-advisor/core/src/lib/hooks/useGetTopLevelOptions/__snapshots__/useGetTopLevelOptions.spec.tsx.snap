// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetTopLevelOptions > should render successfully 1`] = `
{
  "contactLink": "https://www.bambu.life",
  "incomeThreshold": 100000,
  "retireeSavingsThreshold": 100000,
}
`;
